### PR TITLE
chore: upgrade to node version 20 (FE-6252)

### PR DESCRIPTION
**Pull request by Parsium**
### Proposed behaviour

- Upgrade Carbon packages, CI config and scripts to support Node 20 - the latest LTS version.
- Actively tests components on node 18
- No longer actively testing components on node 16
- Regenerate `package-lock.json` to use the latest file format ([more details in the npm docs](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json?v=true#lockfileversion))

### Current behaviour

- Carbon currently supports Node 18 - which is now [no longer the active LTS](https://nodejs.org/en/about/previous-releases).
- Carbon actively tests against node 16

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Playwright automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Related docs have been updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

To assist with regenerating the package lock file, several packages have been locked to specific versions/ restricted to specific version ranges. These packages and their locked versions are below:

   - `@playwright/experimental-core-react17` to `~1.36.2` 
   - `@playwright/test` to `~1.36.2`
   - `@axe-core/playwright` to `~4.7.3`
   - `@sage/design-tokens` (dev dependency only) to `~4.17.0`
   - `@floating-ui/dom` to `~1.2.9`
   - `@floating-ui/react-dom` to `~1.3.0`
   - `cypress` to `~12.13.0`
   - `@types/styled-components` to `5.1.17`[^1]
   - `@tanstack/react-virtual` to `3.0.0-beta.54`

These will eventually need updating to later versions.

[^1]: Should be `4.4.x` to match `styled-components@4.4.1`, but type errors were encountered on the latest `4.4.x` version.

##### Pre-setup
1. Install latest Node LTS using nvm: `nvm install lts/iron`
2. Run `npm ci`

##### Tests
- [ ] Ensure no regressions with `npm` scripts
- [ ] Ensure no regressions with final Carbon build
- [ ] Ensure no regressions with GitHub Actions
- [x] Ensure the Codesandbox CI successfully builds sandbox

### CODE REVIEW

**Comment by Parsium** [comment]
For more detailed context on the new package restrictions, the following packages have been restricted:

| Package                     |       Version       | Reasoning |
| --------------------------- | :-: | ---|
| `@types/styled-components` | 5.1.17        | Ideally should be version 4.x to match `styled-components`, but [unable to downgrade due to found bug](https://github.com/microsoft/TypeScript/issues/37597#issuecomment-628149946). <br/>Versions above 5.1.17 also cause errors due to polymorphic nature of `Button`. Ideal fix is to eventually deprecate rendering `Button` as a link.  |
| `@tanstack/react-virtual`   | 3.0.0-beta.54 | |
| `@floating-ui/dom`          | 1.2.x         | |
| `@floating-ui/react-dom`    | 1.3.x         | |
| `@playwright/experimental-ct-react17`    | 1.40.x         | Peer dependencies were modified in version 1.38, which affected types we've imported. It appears the Playwright team do not use semantic versioning so locked semver range to patches only.  |

File: .codesandbox/ci.json [hunk 1]
```diff
@@ -4,5 +4,5 @@
     "carbon-quickstart-typescript-6sjfuz"
   ],
   "buildCommand": "precompile",
-  "node": "16"
+  "node": "18"
```
Comment (Parsium) [inline .codesandbox/ci.json:7]: **note**: Codesandbox CI currently doesn't support Node 20, [Node 18 is the latest version it supports](https://codesandbox.io/docs/learn/ci#configuration-format)

File: package.json [hunk 2]
```diff
@@ -72,7 +73,8 @@
     "@commitlint/cli": "^17.6.3",
     "@commitlint/config-conventional": "^17.6.3",
     "@dlgshi/cypress-plugin-designtokens": "^0.0.1-development",
-    "@playwright/experimental-ct-react17": "^1.36.1",
+    "@playwright/experimental-ct-react17": "~1.40.0",
+    "@playwright/test": "~1.40.0",
```
Comment (Parsium) [inline package.json:77]: **note**: Playwright has changed the peer dependencies of the component testing library in version 1.38, so we need to install `@playwright/test` separately.

File: .codesandbox/ci.json (hunk 1 above)
Comment (edleeks87) [reply .codesandbox/ci.json:7]: Yeah they're usually a bit behind on updating (understandably)

File: package.json [hunk 3]
```diff
@@ -52,28 +52,30 @@
   },
   "homepage": "https://carbon.sage.com",
   "peerDependencies": {
-    "@sage/design-tokens": "^4.17.0",
+    "@sage/design-tokens": "^4.31.0",
```
Comment (edleeks87) [inline package.json:55]: comment: good idea to update this, let's update the [guidance](https://github.com/Sage/carbon/blob/36093a5b04f82ed9eafedbfd40d07e6f0572aa9e/docs/installation.stories.mdx#L26) to reflect it as well.

File: docs/dev-environment-setup.md [hunk 4]
```diff
@@ -35,11 +35,19 @@ Below are some useful extensions that we recommend using with Carbon:
 
 ### Node.js & NPM
 
-The recommended way to install `node` and `npm` is using [Node Version Manager (`nvm`)](https://github.com/nvm-sh/nvm).
-Once you have installed `nvm`, you should run `nvm install lts/hydrogen` which will install a suitable version of `node` and `npm`.
+The recommended way to install Node.js and npm is using [Node Version Manager (`nvm`)](https://github.com/nvm-sh/nvm).
```
Comment (edleeks87) [inline docs/dev-environment-setup.md:38]: Question: as we're enforcing that anyone developing in Carbon uses node 20 in the `check_node_version` script maybe we should add something about that here?

File: scripts/check_node_version.mjs [hunk 5]
```diff
@@ -1,8 +1,8 @@
-const semver = require("semver");
-const chalk = require("chalk");
+import semver from "semver";
+import chalk from "chalk";
 
-const version = ">=16";
-if (!semver.satisfies(process.version, version)) {
+const versionRange = ">=20.9.0";
```
Comment (edleeks87) [inline scripts/check_node_version.mjs:4]:
```suggestion
const versionRange = ">=20";
```

File: package.json (hunk 3 above)
Comment (Parsium) [reply package.json:55]: Good point, though will consumers get the latest minor version of `@sage/design-tokens` anyway?

Comment (edleeks87) [reply package.json:55]: They will but we may as well update it as we've had people point it out recently

File: playwright/components/anchor-navigation/index.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { Page } from "playwright-core";
+import { Page } from "@playwright/test";
```
Comment (Parsium) [inline playwright/components/anchor-navigation/index.ts:1]: **note**: for future Playwright migrations, we should import any needed TypeScript types from this package as opposed to `playwright-core`, based from recommendation from the Playwright team.

**Review by DipperTheDan — DISMISSED** [review]

**Review by divyajindel — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Pull request merged**